src/curl.c

```diff
@@ -539,7 +526,107 @@ static int swupd_curl_get_file_full(const char *url, char *filename,
 		unlink(filename);
 	}
 
-	return err;
+	return -status;
+}
+
+/*
+ * Determines what strategy to use based on the return code:
+ * - do not retry: if the fault indicates that the failure isn't transient or
+ *                 is unlikely to be successful if repeated (i.e. disk full)
+ * - retry immediately: if the specific fault reported is unusual or rare, it
+ *                      might have been caused by unusual circumstances, the
+ *                      same failure is unlikely to be repeated (e.g. corrupt
+ *                      network package)
+ * - retry after a delay: if the fault is caused by a transient fault (like
+ *                        network connectivity)
+ */
+static enum retry_strategy determine_strategy(int ret_code)
+{
+	switch (ret_code) {
+	case DOWNLOAD_STATUS_FORBIDDEN:
+	case DOWNLOAD_STATUS_NOT_FOUND:
+	case DOWNLOAD_STATUS_WRITE_ERROR:
+		return DONT_RETRY;
+	case DOWNLOAD_STATUS_RANGE_ERROR:
+	case DOWNLOAD_STATUS_PARTIAL_FILE:
+		return RETRY_NOW;
+	case DOWNLOAD_STATUS_ERROR:
+	case DOWNLOAD_STATUS_TIMEOUT:
+		return RETRY_WITH_DELAY;
+	default:
+		return RETRY_NOW;
+	}
+}
+
+static int retry_download(const char *url, char *filename, struct curl_file_data *in_memory_file, bool resume_ok)
+{
+
+	int current_retry = 0;
+	int sleep_time = retry_delay;
+	int strategy;
+	int ret;
+
+	for (;;) {
+
+		/* download file */
+		ret = swupd_curl_get_file_full(url, filename, in_memory_file, resume_ok);
+
+		if (ret == DOWNLOAD_STATUS_COMPLETED) {
+			/* operation successful */
+			break;
+		} else {
+			/* operation failed, determine retry strategy (only if the content
+			 * URL is not local). The value of ret on error will be < 0 */
+			current_retry++;
+			if (content_url_is_local) {
+				strategy = DONT_RETRY;
+			} else {
+				strategy = determine_strategy(-ret);
+			}
+			switch (strategy) {
+			case DONT_RETRY:
+				/* don't retry just return the failure */
+				return ret;
+			case RETRY_NOW:
+				/* if we have reached the retry limit just return the failure,
+				 * if not try again immediately */
+				if (max_retries) {
+					if (current_retry <= max_retries) {
```

Comment:
current_retry < max_retries.

current_retry starts with 0. So in the first retry it will be 1. if max_retries is 1 we shouldn't retry.

Reply:
current_retry is being pre incremented before getting to that if, so it is really as if it started at 1, not 0, so if the user set up a max_retry of 1, if there is a download failure it will retry once and give up. The functional tests added verify the correct behavior of this check them out.

Reply:
Sorry, I didn't notice that. You're right.